examples_tests/48.ArithmeticUnitTest/main.cpp

```diff
@@ -0,0 +1,230 @@
+	auto downloadFence = driver->placeFence(true);
+	auto result = downloadFence->waitCPU(timeoutInNanoSeconds, true);
+	if (result == video::E_DRIVER_FENCE_RETVAL::EDFR_TIMEOUT_EXPIRED || result == video::E_DRIVER_FENCE_RETVAL::EDFR_FAIL)
+	{
+		os::Printer::log("Could not download the buffer from the GPU, fence not signalled!", ELL_ERROR);
+		downloadStagingArea->multi_free(1u, &address, &BUFFER_SIZE, nullptr);
+		return false;
+	}
+	if (downloadStagingArea->needsManualFlushOrInvalidate())
+		driver->invalidateMappedMemoryRanges({ {downloadStagingArea->getBuffer()->getBoundMemory(),address,BUFFER_SIZE} });
+
+	auto dataFromBuffer = reinterpret_cast<uint32_t*>(reinterpret_cast<uint8_t*>(downloadStagingArea->getBufferPointer()) + address);
+	//now check if the data obtained has valid values
+	uint32_t* end = dataFromBuffer + BUFFER_SIZE / sizeof(uint32_t);
+	uint32_t invocationIndex = 0;
+	for (uint32_t* i = dataFromBuffer; i < end; i++)
+	{
+		uint32_t val = *(inputData + invocationIndex);	//TODO instead the template function and compare result
+		/*if (val != *i)
+			goto validationEndFalse;*/
+		invocationIndex++;
+	}
+	downloadStagingArea->multi_free(1u, &address, &BUFFER_SIZE, nullptr);
+	return true;
+
+validationEndFalse:
+	downloadStagingArea->multi_free(1u, &address, &BUFFER_SIZE, nullptr);
+	return false;
+
+}
+int main()
+{
+	// create device with full flexibility over creation parameters
+	// you can add more parameters if desired, check irr::SIrrlichtCreationParameters
+	irr::SIrrlichtCreationParameters params;
+	params.Bits = 24; //may have to set to 32bit for some platforms
+	params.ZBufferBits = 24; //we'd like 32bit here
+	params.DriverType = video::EDT_OPENGL; //! Only Well functioning driver, software renderer left for sake of 2D image drawing
+	params.WindowSize = dimension2d<uint32_t>(1280, 720);
+	params.Fullscreen = false;
+	params.Vsync = true; //! If supported by target platform
+	params.Doublebuffer = true;
+	params.Stencilbuffer = false; //! This will not even be a choice soon
+	params.StreamingDownloadBufferSize = BUFFER_SIZE * 2;
+	auto device = createDeviceEx(params);
+
+	if (!device)
+		return 1; // could not create selected driver.
+
+	video::IVideoDriver* driver = device->getVideoDriver();
+	io::IFileSystem* filesystem = device->getFileSystem();
+	asset::IAssetManager* am = device->getAssetManager();
+
+	uint32_t* inputData = new uint32_t[BUFFER_SIZE / sizeof(uint32_t)];
+
+	auto inputDataBuffer = make_smart_refctd_ptr<ICPUBuffer>(BUFFER_SIZE);
+	{
+		uint32_t* ptr = static_cast<uint32_t*>(inputDataBuffer->getPointer());
+		for (size_t i = 0; i < BUFFER_SIZE / sizeof(uint32_t); i++)
+		{
+			auto memAddr = ptr + i;
+			uint32_t randomValue = std::rand();
+			*memAddr = randomValue;
+			inputData[i] = randomValue;
+		}
+	}
+	core::smart_refctd_ptr<IGPUBuffer> gpuinputDataBuffer = core::smart_refctd_ptr<IGPUBuffer>(driver->getGPUObjectsFromAssets(&inputDataBuffer, &inputDataBuffer + 1)->front()->getBuffer());
```

Comment:
dont do the long way of creating an ICPUBuffer from the array, and then converting it... there's a special function in the driver called "create Filled Device Local On Ded Mem" IIRC which only takes buffer size and contents as parameter